Joy-proposals - extract potentially reusable code into joy-utils

Lezek123 opened the pull request:
This PR contains the initial progress I've made on extracting potentially reusable code from Pioneer packages into `joy-utils` (which may then become part of `joystream-js`). In this particular one I was only extracting code from `joy-proposals` package.

**Rationale:**

The issue was origianlly raised by @Gamaranto here: https://github.com/Joystream/joystream/issues/402

Pioneer consists of many "uncomsumable" packages like `joy-media`, `joy-members` etc., which circulary depend on each other and are pretty much ususable beyond Pioneer, even though they contain a lot of code that could be potentially be reusable.

Among other problems, those packages also usually duplicate a lot of other packages' logic, ie. a few of them (`joy-media`, `joy-roles`, `joy-proposals`) define their own transport layer with methods like `blockTimestamp` which returns timestamp by a given block number (a method that could obviously be useful for other packages/projects). There are also a lot of hooks, components, functions, constants etc. like that, which we currently either import between packages causing circular dependency hell or rewrite in the scope of each new package, losing time and multiplying code and potential points of failure.

**The new structure of `joy-utils`**

In order to prepare `joy-utils` to be an independent library (or part of such) that could handle all this reusable Pioneer's code I ended up with 5 base categories of code that we could put there:
- `consts` - there is a noticable amount of hardcoded constants that cannot either yet or ever be fetched from the runtime. In `joy-proposals` those are for example: validation constraints (enforced by runtime), proposal types, approval/slashing quorum and threshold by type, required stakes by type etc. In `joy-media` we have schemas like `Video`, `Channel`, `Content` etc. that would also be more useful as part of a separate library. By extracting those out we could share them among other projects and not have to hardcode them multiple times.
- `functions` - this group will contain all the functions defined in the Pioneer packages that could be reused (either by other packages or projects), ie. some common calculations, converters etc.
- `transport` - the aim here is to create a unified interface to fetch data from the api. This is probably the hardest one to get right, because it would need to also handle mock transports, potential caching, subscriptions etc. Having a unified transport layer we would potentially also need to re-think the logic of HOCs like `withMyAccount` etc. (I don't think this needs to be done all at once though, besides it should be considered how query node and other WIP solutions could affect this). My current idea is that there will be one global `Transport` instance for the Pioneer (available through context) that will internally instantialize more specific transports like `ProposalsTransport` or `MembersTransport`. Those more specific transports could be injected to each other in the main `Transport` constructor (ie. `CouncilTransport` could have `MembersTransport` instance injected). This way we will have a very clear chain of dependency and avoid getting back to circular references, while maintaining some kind of modularity. Inside Pioneer, for example, we could use `transport.proposals.all()` to fetch all proposals (with additional information collected by calling various lower-level api methods) or `transport.council.activeMembers()` to fetch the profiles of all active council members. Having Pioneer rebuild in a way that it will rely on a unified transport layer should allow us to switch the actual transport implementation more easily when needed.
- `react` - here we could keep reusable code specific to react, like `hooks`, `components`, `contexts` etc. This could then be extracted as a separate library, having the actual `joystream-js` just as a dependency.
- `types` - here we could keep types that don't need to be part of `@joystream/types` but will be still valueable for some higher-level interactions with the data. For example, in case of `joy-proposals` we have `ParsedProposal` which contains a creation date (based on creation block), fetched and parsed creator profile etc. (although this type itself may need some improvements, it's way easier to use a construct like that in order to display the data on the frontend instead of operating just on the types available in `@joystream/types`).

**The scope of this PR**

In this specific PR I was tyring to re-build `joy-proposals`. having the above structure in mind. It was mostly about splitting the code and putting it in the right places inside the structure I created in `joy-utils`, I wasn't trying to change the actual implementation too much. Obviously this will not allow us to see the real benefits of refactoring the entire Pioneer this way, but it may server as a good starting point for a discussion and allow us to test some implications of such a rebuild on a smaller scale first.

Commit:
joy-proposals - extract potentially reusable code into joy-utils

Commit:
Remove needless files after rebase

Commit:
Merge branch 'development' into joy-utils-rework

fcescob commented on pioneer/packages/joy-utils/src/functions/misc.ts:
```diff
@@ -0,0 +1,32 @@
+  return str.split(/(?=[A-Z])/);
+}
+
+export function slugify(str: string) {
```
If our future plan is to make `joy-utils` a consumable module, we should strive to include only the core necessities to both reduce bundle size and allow for easier documentation.
I would keep here only the functions used internally by `joy-utils` and leave the others implementation to the consumers of `joy-utils`.

fcescob commented on pioneer/packages/joy-utils/src/functions/proposals.ts:
```diff
@@ -0,0 +1,152 @@
+}
+
+export function calculateMetaFromType(type: ProposalType): ProposalMeta {
+  const image = "";
```
`image` here is about the UI, I would leave this to the consumers to deal with the it as well.

fcescob reviewed (commented):
Left two small comments, but looks very good to me, I really like the approach you took with the transports. Great job!

Commit:
Minor adjustments

Lezek123 commented:
Added a few minor adjustments that should fix the mentioned issues:
- Moved hardcoded proposals data from `functions/proposals` to `consts/proposals` (as that's where it really belongs) and changed the structure of it from functions to a single object with proposal types as keys. Figured out we don't need to separate stakes from other hardcoded data, so just added it as part of `ProposalMeta`.
- Removed functions like `snakeCaseToCamelCase`, `slugify` etc. and replaced them with `loadash` equivalents.
- Removed `image` as part of hardcoded `ProposalMeta` (we don't have this implemented yet anyway)
